src/coreclr/tools/Common/Pgo/PgoFormat.cs

```diff
@@ -0,0 +1,520 @@
+                if ((modifyMask & InstrumentationDataProcessingState.Count) == InstrumentationDataProcessingState.Count)
+                    valueEmitter.EmitLong(schema.Count, prevSchema.Count);
+                if ((modifyMask & InstrumentationDataProcessingState.Other) == InstrumentationDataProcessingState.Other)
+                    valueEmitter.EmitLong(schema.Other, prevSchema.Other);
+
+                for (int i = 0; i < schema.Count; i++)
+                {
+                    switch (schema.InstrumentationKind & PgoInstrumentationKind.MarshalMask)
+                    {
+                        case PgoInstrumentationKind.None:
+                            break;
+                        case PgoInstrumentationKind.FourByte:
+                        {
+                            long valueToEmit;
+                            if (schema.Count == 1)
+                            {
+                                valueToEmit = schema.DataLong;
+                            }
+                            else
+                            {
+                                valueToEmit = ((int[])schema.DataObject)[i];
+                            }
+                            valueEmitter.EmitLong(valueToEmit, prevEmittedIntData);
+                            prevEmittedIntData = valueToEmit;
+                            break;
+                        }
+                        case PgoInstrumentationKind.EightByte:
+                        {
+                            long valueToEmit;
+                            if (schema.Count == 1)
+                            {
+                                valueToEmit = schema.DataLong;
+                            }
+                            else
+                            {
+                                valueToEmit = ((long[])schema.DataObject)[i];
+                            }
+                            valueEmitter.EmitLong(valueToEmit, prevEmittedIntData);
+                            prevEmittedIntData = valueToEmit;
+                            break;
+                        }
+                        case PgoInstrumentationKind.TypeHandle:
+                        {
+                            TType typeToEmit = ((TType[])schema.DataObject)[i];
+                            valueEmitter.EmitType(typeToEmit, prevEmittedType);
+                            prevEmittedType = typeToEmit;
+                            break;
+                        }
+                    }
+                }
+
+                prevSchema = schema;
+            }
+
+            // Emit a "done" schema
+            valueEmitter.EmitDone();
+        }
+
+        public static PgoSchemaElem[] Merge<TType>(ReadOnlySpan<PgoSchemaElem[]> schemasToMerge)
+        {
+            // The merging algorithm will sort the schema data by iloffset, then by InstrumentationKind
+            // From there each individual instrumentation kind shall have a specific merging rule
```

Comment:
For edge-based profiling (#46882) I need to identify edges and was going to use IL pairs for this.

I was thinking of using `Other` to hold the second IL offset, which would mean for some record types `Other` must be part of the sort key, but for other types it should not be, or more complex merging is needed to search among all the records with the same IL offset.

If necessary I can find other ways to encode edges so there is a 32 bit dense representation, but it would be nice not to have to resort to this.

Reply:
Hm, Ok, I remember thinking that Other would be good for things like that, and this would preclude such simple plans. Hmm... I'll have to think on that for a bit.

Reply:
Merging already has to be aware of the various instrumentation kinds in order to correctly operate on the data portions of the stream. As such the merging algorithm can be made aware of whether or not the Other field should be considered part of the "sortkey" here. Would that alleviate your concerns?

Reply:
Also, if we want to, the Other and ILOffset fields could be expanded to 64bits without changing any of the long-term storage formats.

Reply:
Having Other be conditionally meaningful for sorting would be preferable.